### ISSUE TITLE
Undefined Overflow Behaviour in Calls to cStructGenLakes

### ISSUE THREAD

**Issue by worktycho** [mc-server/MCServer#1797]
Seed value can be large so can cause overflows in the multiplication that is passed the constructor. This should be fixed to avoid the overflow. Relevant error from clang sanatize run:

> /home/tommy/MCServer/src/Generating/ComposableGenerator.cpp:357:66: runtime error: signed integer overflow: 660851791 * 5 cannot be represented in type 'int'

**Comment by gauraVsehgaL** [comment]
I see the following code at Line at 371.
`m_FinishGens.push_back(cFinishGenPtr(new cStructGenLakes(Seed * 5 + 16873, E_BLOCK_STATIONARY_LAVA, m_ShapeGen, Probability)));`

So I guess the problem is in the above line.Can we modulo `Seed * 5 + 16873` value so that it fits into an integer. Something like `(Seed * 5 + 16873)%INT_MAX`.
Similar problem may arise at line 561
`m_FinishGens.push_back(cFinishGenPtr(new cStructGenLakes(Seed * 3 + 652, E_BLOCK_STATIONARY_WATER, m_ShapeGen, Probability)));`

**Comment by gauraVsehgaL** [comment]
Is this still open?

**Comment by madmaxoft** [comment]
There are way too many places where this behavior is used. And none of them really matter, since the numbers are used only as random numbers, so it doesn't matter much if they are clobbered in an overflow.

Is there even a sane way to tell C++ that's what we want?

**Comment by tigerw** [comment]
No. You will follow that standard, or face its wrath. All >1000 pages of it, *at speed*.

And if that doesn't quail you, you'll also have do deal with @worktycho. Good luck!

**Comment by satoshinm** [comment]
Would it be possible to switch these integers to unsigned? (since unsigned overflow is defined, unlike signed overflow which C leaves undefined, so the result could differ on other platforms)

**Comment by madmaxoft** [comment]
I suppose it could be done, but it has to be done extra carefully - you don't want to change the values that got calculated, otherwise you change world generation for existing servers.

**Comment by pwnorbitals** [comment]
Is this really a problem ? Do we want to guarantee generation stability across major releases ?

**Comment by tigerw** [comment]
The undefined behaviour might be a problem technically but probably (?) isn't anything serious.
I'm not sure about stability, I believe xoft wanted to maintain some degree of it (i.e. no hard chunk borders).
